indexing-service/src/main/java/io/druid/indexing/common/actions/SegmentTransactionalInsertAction.java

```diff
@@ -99,14 +99,24 @@ public DataSourceMetadata getEndMetadata()
   @Override
   public SegmentPublishResult perform(Task task, TaskActionToolbox toolbox) throws IOException
   {
-    toolbox.verifyTaskLocks(task, segments);
+    for (DataSegment segment : segments) {
+      if (!toolbox.getTaskLockbox().upgrade(task, segment.getInterval()).isOk()) {
+        return SegmentPublishResult.fail();
```

Comment:
I guess the fail ```SegmentPublishResult``` is not common because it can happen only when the lock upgrade is failed due to preemption or metadata update is failed. In this case, I think this is better because the instance will be GCed immediately after it is processed while the static instance will be kept until the jvm terminates.